File: backend/src/models/bulk_color_approval.rs
//! 大货批色审批 Model（V15 P0-F15 创建）
//!
//! 表 bulk_color_approval：记录面料大货的批色流程
//! 业务：剪样 → 发送客户 → 客户批色确认 → 通过/拒绝/返工/降级/报废
//!
//! 8 态状态机：
//!   pending → sampled → sent_to_customer → approved / rejected / rework
//!                                                    ↓
//!                                               downgraded / scrapped
//!
//! 关联任务：P0-F15（建表）/ P0-F16（剪大货样）/ P0-F17（客户批色确认）/ P0-F19（ship_order 校验）

use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

/// 大货批色审批实体
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "bulk_color_approval")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i64,
    /// 关联销售订单
    pub sales_order_id: i32,
    /// 关联染色批次（缸号）
    pub dye_batch_id: i32,
    /// 客户 ID
    pub customer_id: i64,
    /// 关联生产订单（可选，剪样时填入）
    pub production_order_id: Option<i32>,
    /// 四维标识：产品 ID
    pub product_id: Option<i32>,
    /// 四维标识：色号
    pub color_no: Option<String>,
    /// 四维标识：染色批号（lot 概念，防色差混批）
    pub dye_lot_no: Option<String>,
    /// 四维标识：批次号（缸号）
    pub batch_no: Option<String>,
    /// 样布类型：cut_sample(剪大货样) / lab_sample(化验室打样)
    pub sample_type: String,
    /// 样布 inventory_piece ID（剪样扣减后产生的 sample piece）
    pub sample_piece_id: Option<i64>,
    /// 样布长度（米）
    pub sample_length_m: Option<Decimal>,
    /// 批色状态：pending/sampled/sent_to_customer/approved/rejected/rework/downgraded/scrapped
    pub approval_status: String,
    /// 审批人用户 ID
    pub approver_id: Option<i32>,
    /// 批准时间
    pub approval_date: Option<DateTime<Utc>>,
    /// 发送给客户时间（批色时限计算锚点）
    pub sent_to_customer_at: Option<DateTime<Utc>>,
    /// 客户反馈
    pub customer_feedback: Option<String>,
    /// CIE D65 色差值 ΔE
    pub delta_e_value: Option<Decimal>,
    /// 拒绝/返工原因
    pub reject_reason: Option<String>,
    /// 交货门禁标志（true 时阻止发货）
    pub delivery_blocking: bool,
    /// 附件 URL（样布照片/批色单扫描件）
    pub attachment_url: Option<String>,
    /// 备注
    pub remark: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// 关联关系
#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    /// 批色记录 - 销售订单（多对一）
    #[sea_orm(
        belongs_to = "super::sales_order::Entity",
        from = "Column::SalesOrderId",
        to = "super::sales_order::Column::Id"
    )]
    SalesOrder,
    /// 批色记录 - 染色批次（多对一）
    #[sea_orm(
        belongs_to = "super::dye_batch::Entity",
        from = "Column::DyeBatchId",
        to = "super::dye_batch::Column::Id"
    )]
    DyeBatch,
    /// 批色记录 - 客户（多对一）
    #[sea_orm(
        belongs_to = "super::customer::Entity",
        from = "Column::CustomerId",
        to = "super::customer::Column::Id"
    )]
    Customer,
}

impl Related<super::sales_order::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::SalesOrder.def()
    }
}

impl Related<super::dye_batch::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::DyeBatch.def()
    }
}

impl Related<super::customer::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Customer.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}

File: frontend/src/api/bulk-color-approval.ts
import { request } from './request';
import type { ApiResponse, PaginatedResponse } from '@/types/api';

/**
 * 大货批色审批（bulk_color_approval）状态
 * 后端状态机：pending→sampled→sent_to_customer→approved/rejected/rework；downgraded/scrapped 为终态
 */
export type BulkColorApprovalStatus =
  | 'pending'
  | 'sampled'
  | 'sent_to_customer'
  | 'approved'
  | 'rejected'
  | 'rework'
  | 'downgraded'
  | 'scrapped';

export interface BulkColorApproval {
  id: number;
  sales_order_id: number;
  dye_batch_id: number;
  customer_id: number;
  production_order_id: number | null;
  product_id: number | null;
  color_no: string | null;
  dye_lot_no: string | null;
  batch_no: string | null;
  sample_type: string;
  sample_length_m: string | number | null;
  approval_status: BulkColorApprovalStatus;
  approver_id: number | null;
  approval_date: string | null;
  sent_to_customer_at: string | null;
  customer_feedback: string | null;
  delta_e_value: string | number | null;
  reject_reason: string | null;
  delivery_blocking: boolean;
  attachment_url: string | null;
  remark: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * 创建大货批色申请（对齐后端 CreateBulkColorApprovalDto，bulk_color_approval_handler.rs:44）。
 * 大货样确认必须绑定销售订单+染缸批次：sales_order_id、dye_batch_id、customer_id 均必填；其余可选。
 */
export interface CreateBulkColorApprovalPayload {
  sales_order_id: number;
  dye_batch_id: number;
  customer_id: number;
  production_order_id?: number;
  product_id?: number;
  color_no?: string;
  dye_lot_no?: string;
  batch_no?: string;
  sample_type?: string;
  remark?: string;
}

export interface CutSamplePayload {
  sample_length_m: number;
  sample_piece_id?: number;
  attachment_url?: string;
  delta_e_value?: number;
}

/**
 * 大货批色审批列表查询参数——严格对齐后端
 * bulk_color_approval_service.rs::ListBulkColorApprovalQuery。
 * 全部 Option 字段 → 可选；无 rename_all → 保持 snake_case。
 * 状态字段为 approval_status（不是 status），日期为 from_date/to_date（ISO 串）。
 */
export interface BulkColorApprovalListQuery {
  page?: number;
  page_size?: number;
  sales_order_id?: number;
  dye_batch_id?: number;
  customer_id?: number;
  approval_status?: string;
  from_date?: string;
  to_date?: string;
}

export function getBulkColorApprovalList(
  params?: BulkColorApprovalListQuery
): Promise<ApiResponse<PaginatedResponse<BulkColorApproval>>> {
  return request.get('/bulk-color-approvals', { params });
}

export function getBulkColorApproval(id: number): Promise<ApiResponse<BulkColorApproval>> {
  return request.get(`/bulk-color-approvals/${id}`);
}

export function createBulkColorApproval(
  data: CreateBulkColorApprovalPayload
): Promise<ApiResponse<BulkColorApproval>> {
  return request.post('/bulk-color-approvals', data);
}

export function cutBulkColorSample(
  id: number,
  data: CutSamplePayload
): Promise<ApiResponse<BulkColorApproval>> {
  return request.post(`/bulk-color-approvals/${id}/cut-sample`, data);
}

export function sendBulkColorToCustomer(id: number): Promise<ApiResponse<BulkColorApproval>> {
  return request.post(`/bulk-color-approvals/${id}/send-to-customer`);
}

export function approveBulkColor(
  id: number,
  data?: { feedback?: string; delta_e_value?: number }
): Promise<ApiResponse<BulkColorApproval>> {
  return request.post(`/bulk-color-approvals/${id}/approve`, data ?? {});
}

export function rejectBulkColor(
  id: number,
  data: { reject_reason: string; feedback?: string }
): Promise<ApiResponse<BulkColorApproval>> {
  return request.post(`/bulk-color-approvals/${id}/reject`, data);
}

export function reworkBulkColor(
  id: number,
  data: { reject_reason: string; feedback?: string }
): Promise<ApiResponse<BulkColorApproval>> {
  return request.post(`/bulk-color-approvals/${id}/rework`, data);
}

export function downgradeBulkColor(
  id: number,
  data: { reject_reason: string }
): Promise<ApiResponse<BulkColorApproval>> {
  return request.post(`/bulk-color-approvals/${id}/downgrade`, data);
}

export function scrapBulkColor(
  id: number,
  data: { reject_reason: string }
): Promise<ApiResponse<BulkColorApproval>> {
  return request.post(`/bulk-color-approvals/${id}/scrap`, data);
}

export function getBulkColorApprovalHistory(id: number): Promise<ApiResponse<unknown[]>> {
  return request.get(`/bulk-color-approvals/${id}/history`);
}

export function getBulkColorApprovalStatistics(): Promise<ApiResponse<unknown>> {
  return request.get('/bulk-color-approvals/statistics');
}
